### PR TITLE
Reduce the size of System.Net.IPAddress

### PR DESCRIPTION
**Pull request by justinvp**
This change reduces the size of `IPAddress` from 40 bytes to 24 bytes (64 bit) and also avoids an unnecessary array allocation for IPv4 addresses.

cc: @davidsh, @CIPop, @stephentoub

These are the current fields on `IPAddress` that add up to 40 bytes (64 bit):

```c#
internal long Address;
internal string StringRepresentation;
private readonly AddressFamily _family = AddressFamily.InterNetwork;
private readonly ushort[] _numbers = new ushort[NumberOfLabels];
private long _scopeId = 0;
private int _hashCode = 0;
```

 - `Address` is only used for IPv4 addresses. The size is actually `uint`, but it is typed as `long` because public APIs use `long` for CLS compliance. This field can be changed to `uint` to save 4 bytes.

 - `StringRepresentation` is a cache of the result of `ToString()` and can remain as-is.

 - `_family` indicates whether the `IPAddress` instance is an IPv4 address or an IPv6 address through the `AddressFamily.InterNetwork` (IPv4) and `AddressFamily.InterNetworkV6` (IPv6) enum values. We can eliminate this field and instead use the `_numbers` field, which is only used for IPv6 addresses, to indicate whether or not the address is IPv4 vs. IPv6. This saves 4 bytes.

 - `_numbers`, as mentioned in the previous bullet, is only used for IPv6 addresses. Note that currently the array is always being allocated, even for IPv4 addresses. We can change this to only allocate the array for IPv6 addresses. When the field is null, it indicates that the address is an IPv4 address, when the field is non-null it is an IPv6 address.

 - `_scopeId` is only used for IPv6 addresses. Similar to `Address`, the size is actually `uint`, but it is typed as `long` because public APIs use `long` for CLS compliance. We can eliminate this field and instead use the same `uint` field as `Address` for IPv4 addresses, reducing the size by another 8 bytes.

 - `_hashCode` is only used for IPv6 addresses (IPv4 just uses the `Address` for the hash code). We could consider getting rid of this field, to reduce the size by another 4 bytes, and instead always re-calculate the hash code when `GetHashCode` is called (all it's doing is calling `StringComparer.OrdinalIgnoreCase.GetHashCode()` on a cached `ToString()` result), but I assume it is here for performance reasons, so I've kept it for now.

Here are the fields after the changes:

```C#
private uint _addressOrScopeId;
private readonly ushort[] _numbers;
private string _toString;
private int _hashCode;
```

From 40 bytes to 24 bytes (64 bit).

(Note: I renamed some of the fields and made them all private. The two renamed internal fields (`Address` and `StringRepresentation`) now have internal properties with the same type and name as the old fields, to avoid breaking any internal callers.)

(There are other potential optimizations for `IPAddress` that could be handled in separate PRs. For example, `ToString()` could be optimized to avoid unnecessary intermediate allocations. Also, as noted in #2891, some means to access to the internal buffer to avoid unnecessary `byte[]` allocations could be added.)

### CODE REVIEW

File: src/System.Net.Primitives/src/System/Net/IPAddress.cs [hunk 1]
```diff
@@ -56,6 +59,29 @@ public class IPAddress
...
 
+        // When IPv6 support was added to the .NET Framework, the public Address property was marked as Obsolete.
+        // The public obsolete Address property has not been carried forward in .NET Core, but remains here as
+        // internal so internal types that understand IPv4 can still access it without obsolete warnings.
+        internal long Address
+        {
+            get { return _addressOrScopeId; }
+        }
+
+        internal string StringRepresentation
```
Comment (justinvp) [inline src/System.Net.Primitives/src/System/Net/IPAddress.cs]: I couldn't find any internal callers of `StringRepresentation` in corefx, but added the property just in case there is other code that isn't open source yet that needs to call this member. Otherwise, this property can go away. (Even if there are internal callers that need it; I'd wonder why they couldn't be updated to call `ToString()` instead)

**Comment by stephentoub** [comment]
@dotnet-bot test this please (https://github.com/dotnet/corefx/issues/4071)

File: src/System.Net.Primitives/src/System/Net/IPAddress.cs [hunk 2]
```diff
@@ -56,6 +59,29 @@ public class IPAddress
         //   1: Terminating null byte
         internal const int NumberOfLabels = IPAddressParserStatics.IPv6AddressBytes / 2;
 
+        // When IPv6 support was added to the .NET Framework, the public Address property was marked as Obsolete.
+        // The public obsolete Address property has not been carried forward in .NET Core, but remains here as
+        // internal so internal types that understand IPv4 can still access it without obsolete warnings.
+        internal long Address
+        {
+            get { return _addressOrScopeId; }
```
Comment (stephentoub) [inline src/System.Net.Primitives/src/System/Net/IPAddress.cs]: Should we assert InternalIsIPv4?  Or are there cases where Address is accessed when IPv6?

File: src/System.Net.Primitives/src/System/Net/IPAddress.cs [hunk 3]
```diff
@@ -148,7 +172,7 @@ public IPAddress(byte[] address)
         // and winsock only understands Int32.
         internal IPAddress(int newAddress)
         {
-            Address = (long)newAddress & 0x00000000FFFFFFFF;
+            _addressOrScopeId = (uint)newAddress & 0x00000000FFFFFFFF;
```
Comment (stephentoub) [inline src/System.Net.Primitives/src/System/Net/IPAddress.cs]: Should we be asserting that newAddress >= 0 and <= uint.MaxValue?  I realize this code is preexisting, but It's not clear to me why the 0xFFFFFFFF is necessary here.  If we're expecting values outside of the uint range, why is it safe to truncate like this, and if we're not expecting such values, why is this needed?

**Comment by stephentoub** [comment]
LGTM, but @davidsh and @cipop should review.

**Comment by CIPop** [comment]
Thanks for the enhancement @justinvp!

To properly validate there is no behavior change we'd need better code coverage tests for the changed code.

The process for refactoring / adding perf improvements would be to:
  1. have pinning behavioral tests for each code path that was affected 
  2. use some code coverage tool to prove that the tests completely cover all cases that you wish to enhance
  3. add the changes in this PR and ensure that the old tests still execute without any changes

File: src/System.Net.Primitives/src/System/Net/IPAddress.cs (hunk 1 above)
Comment (CIPop) [inline src/System.Net.Primitives/src/System/Net/IPAddress.cs]: We don't allow dead code. If there are no callers this must be removed.

File: src/System.Net.Primitives/src/System/Net/IPAddress.cs [hunk 4]
```diff
@@ -21,25 +19,30 @@ public class IPAddress
...
-        ///     Default to IPv4 address
-        ///   </para>
-        /// </devdoc>
-        private readonly AddressFamily _family = AddressFamily.InterNetwork;
-        private readonly ushort[] _numbers = new ushort[NumberOfLabels];
-        private long _scopeId = 0;
-        private int _hashCode = 0;
+        /// <summary>
+        /// For IPv4 addresses, this field stores the Address.
+        /// For IPv6 addresses, this field stores the ScopeId.
```
Comment (CIPop) [inline src/System.Net.Primitives/src/System/Net/IPAddress.cs]: This is a definite no-go. Code clarity and maintainability is more important than the few bytes that you've saved.

There are a lot of other options if you wish to save the bytes:
- create different subclasses for ipv4/v6 (preferred for code clarity)
- use an _union_ style structure (see [FieldOffsetAttribute](https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.fieldoffsetattribute(v=vs.110).aspx))

File: src/System.Net.Primitives/src/System/Net/IPAddress.cs (hunk 3 above)
Comment (sharwell) [inline src/System.Net.Primitives/src/System/Net/IPAddress.cs]: The mask is now just pointless. Its original purpose was to cause `(long)newAddress` to behave as though it were `(long)(uint)newAddress`. In the new code, `(uint)newAddress` is sufficient.

**Comment by sharwell** [comment]
:bulb: I would prefer to see all accesses to `_addressOrScopeId` go through either `Address` or `ScopeId`, and that each of those properties contain an assertion (`Debug.Assert`?) that it is used on an address of the correct form. The performance in release builds would be comparable, but the code would be much easier to understand regarding the intent of each access.

File: src/System.Net.Primitives/src/System/Net/IPAddress.cs (hunk 4 above)
Comment (sharwell) [inline src/System.Net.Primitives/src/System/Net/IPAddress.cs]: :thought_balloon: *As a backing field*, I don't find this difficult to understand. However, as noted in a separate comment the clarity is lost when you access it directly. If all the accesses went through either the `Address` or `ScopeId` property, then I think this approach would be reasonable. I'm not currently convinced that the described alternatives would produce a net advantage in readability over the current approach.

Comment (stephentoub) [inline src/System.Net.Primitives/src/System/Net/IPAddress.cs]: > I'm not currently convinced that the described alternatives would produce a net advantage in readability over the current approach.

And they'd have costs of their own, e.g. virtual calls where none exist today. I like the idea of a backing field accessed via properties with asserts.

File: src/System.Net.Primitives/src/System/Net/IPAddress.cs (hunk 1 above)
Comment (justinvp) [inline src/System.Net.Primitives/src/System/Net/IPAddress.cs]: Removed.

File: src/System.Net.Primitives/src/System/Net/IPAddress.cs (hunk 2 above)
Comment (justinvp) [inline src/System.Net.Primitives/src/System/Net/IPAddress.cs]: This now asserts.

File: src/System.Net.Primitives/src/System/Net/IPAddress.cs (hunk 3 above)
Comment (justinvp) [inline src/System.Net.Primitives/src/System/Net/IPAddress.cs]: I removed the mask, but didn't add any new asserts.

File: src/System.Net.Primitives/src/System/Net/IPAddress.cs (hunk 4 above)
Comment (justinvp) [inline src/System.Net.Primitives/src/System/Net/IPAddress.cs]: I added two private properties (`PrivateAddress` and `PrivateScopeId`) that wrap access to the backing field with asserts.

**Comment by justinvp** [comment]
Thanks for the feedback, everyone. PR updated.

@CIPop, regarding test coverage... The line and branch coverage of `IPAddress` with the existing functional tests was already effectively 100% before these changes (other than two internal members: `Snapshot()` and `Equals(object comparandObj, bool compareScopeId)`), as determined by running the System.Net.Primitives functional tests with `/p:Coverage=true`. I went ahead and augmented the tests with some additional verification. Is there anything else in particular you want to see?

**Comment by CIPop** [comment]
+1 on @sharwell 's idea.

Thanks @justinvp for testing this thoroughly! 

LGTM
